test/test_buildpacks.go

```diff
@@ -0,0 +1,207 @@
+///usr/bin/true; exec /usr/bin/env go run "$0" "$@"
+
+package main
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"os/signal"
+	"path/filepath"
+	"syscall"
+	"time"
+)
+
+const repo = "ghcr.io/boson-project"
+
+func main() {
+	ctx, cancel := context.WithCancel(context.Background())
+	sigs := make(chan os.Signal, 1)
+	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
+
+	go func() {
+		<-sigs
+		cancel()
+		<-sigs
+		os.Exit(1)
+	}()
+
+	version := os.Args[1]
+	err := runTests(ctx, version)
+	if err != nil {
+		fmt.Printf("::error::%s\n", err.Error())
+		os.Exit(1)
+	}
+}
+
+var buildpacks = []struct {
+	Buildpacks []string
+	Runtimes   []string
+}{
+	{
+		Buildpacks: []string{"ghcr.io/boson-project/go-function-buildpack", "paketo-buildpacks/go-dist"},
+		Runtimes:   []string{"go"},
+	},
+	{
+		Buildpacks: []string{"ghcr.io/boson-project/typescript-function-buildpack", "paketo-buildpacks/nodejs"},
+		Runtimes:   []string{"typescript"},
+	},
+}
+
+func runTests(ctx context.Context, version string) error {
+
+	os.Setenv("FUNC_REGISTRY", repo)
+
+	for i := range buildpacks {
+		buildpacks[i].Buildpacks[0] = fmt.Sprintf("%s:%s", buildpacks[i].Buildpacks[0], version)
+	}
+
+	wd, err := os.Getwd()
+	if err != nil {
+		return err
+	}
+
+	packCmd := "pack"
+	if pc, ok := os.LookupEnv("PACK_CMD"); ok {
+		packCmd = pc
+	}
+
+	funcBinaries := []string{
+		filepath.Join(wd, "bin", "func_snapshot"),
+	}
+	templates := []string{
+		"http",
+		"events",
+	}
+
+	oldWD, err := os.Getwd()
+	if err != nil {
+		return err
+	}
+
+	testDir, err := os.MkdirTemp("", "test_func")
+	if err != nil {
+		return err
+	}
+	// defer os.RemoveAll(testDir)
+
+	err = os.Chdir(testDir)
+	if err != nil {
+		return err
+	}
+	defer os.Chdir(oldWD)
+
+	// just a counter to avoid name collision
+	var fnNo int
+
+	for _, funcBinary := range funcBinaries {
+		for _, buildpack := range buildpacks {
+			for _, runtime := range buildpack.Runtimes {
+				for _, template := range templates {
+					fnName := fmt.Sprintf("fn-%s-%s-%d", runtime, template, fnNo)
+					if !runTest(ctx, packCmd, funcBinary, runtime, template, fnName, buildpack.Buildpacks) {
```

Comment:
This is _so_ close to a table-driven test!  It might be worth the effort to go all the way through, unless I am missing an important reason for using the binaries.  Perhaps I will try to write up an exploration of that possibility.